metal/mmtl/slicing/slice_model.py

```diff
@@ -0,0 +1,229 @@
+        * Binary classification heads (with output_dim=1, breaking Metal convention)
+        * A single base task + an arbitrary number of slice tasks (slice_head_type != None)
+    """
+
+    def __init__(self, tasks, **kwargs):
+        base_task = kwargs.get('base_task',None)
+        validate_slice_tasks(tasks, base_task=base_task)
+        super().__init__(tasks, **kwargs)
+        if base_task is None:
+            self.base_task = [
+                t for t in self.task_map.values() if t.slice_head_type is None
+            ][0]
+        else:
+            self.base_task = base_task
+        
+        print(f"Using slicing base task {base_task.name}...")
+
+        self.pred_tasks = {
+           name: t for name, t in self.task_map.items() if t.slice_head_type is None
+        }
+        self.slice_pred_tasks = {
+            name: t for name, t in self.task_map.items() if t.slice_head_type == "pred"
+        }
+        self.slice_ind_tasks = {
+            name: t for name, t in self.task_map.items() if t.slice_head_type == "ind"
+        }
+
+    #def forward_pred(self, X)
+    def forward_body(self, X):
+        """ Makes a forward pass through the "body" of the network
+        (everything before the head)."""
+
+        input = move_to_device(X, self.config["device"])
+        base_task_name = self.base_task.name
+
+        # Extra .module because of DataParallel wrapper!
+        input_module = self.input_modules[base_task_name].module
+        middle_module = self.middle_modules[base_task_name].module
+        attention_module = self.attention_modules[base_task_name].module
+
+        out = attention_module(middle_module(input_module(input)))
+        return out
+
+    def forward_heads(self, body, task_names):
+        """ Given body (everything before head) representation, return dict of
+        task_head outputs for specified task_names """
+
+        return {t: self.head_modules[t].module(body) for t in task_names}
+
+    def forward_attention_logits(self, slice_ind_heads, slice_task_names):
+        """ Computes unnomralized slice_head attention weights based on
+        `in slice` indicators """
+
+        # slice_preds is the [batch_size, num_slices] concatenated prediction for
+        # each slice head
+        slice_inds = torch.cat(
+            [slice_ind_heads[name]["data"] for name in slice_task_names], dim=1
+        )
+
+        # A_weights is the [batch_size, num_slices] unormalized Tensor where
+        # more positive values indicate more likely to be in the slice
+        A_weights = slice_inds
+        return A_weights
+
+    def forward(self, X, task_names):
+        """ Perform forward pass with slice-reweighted base representation through
+        the base_task head. """
+
+        # First, running through the non-sliced tasks
+        pred_names = sorted( 
+            [task_name for task_name in self.pred_tasks.keys()]
+        )  
+
+        pred_heads = super(SliceModel,self).forward(X, pred_names)
```

Comment:
Why do you need to call `super(SliceModel,self).forward`? 
Might be cleaner to use the pattern below:

```
body = self.forward_body(X)
pred_heads = self.forward_heads(body, pred_names)
```

Then, you can share a pass through the body

Reply:
This doesn't work for the multitask case where we need to cache, I don't think.  So I run it once for now through the MTL caching forward pass, and then a second time (if necessary) through the slicing body.  This could be updated.